### DISCUSSION TITLE
PartialForceOutput and StressForceOutput model builders

### DISCUSSION THREAD

**Discussion by youngjae142** [mir-group/nequip#285] [Q&A]
Hello!
I'm curious about the role of the _PartialForceOutput_ and _StressForceOutput_ model builders.

1. Are they the parameters related to LAMMPS?
**2. Does StressForceOutput enable the box and atomic stress(virial) calculations of LAMMPS (pair-nequip)?**
3. What is the meaning of PartialForceOutput and StressForceOutput model builders and what is the difference?
4. Is there any way to find a list of all model builder parameters with short descriptions?

**Comment by Linux-cpp-lisp** [comment]
Hi @ssrokyz ,

In order:

1. Yes, `StressForceOutput` allows you to run NPT simulations and get the _global_ virial/stress/pressure in LAMMPS.  There are prototypes of this in both `pair_nequip` and `pair_allegro`, with the one in `pair_nequip` being more mature.  Please let us know if you find any issues in these implementations as we are trying to complete and merge them into the main releases.  Our implementation does not, however, support _atomic_ virials (which are generally not uniquely defined)---  are these necessary for some application you want to run?
2. `PartialForceOutput` computes partial forces using `torch.autograd.functional` and is neither fast nor TorchScript/LAMMPS compatible.  `StressForceOutput` *is* efficient and TorchScript/LAMMPS compatible, and is the basis of our virial / stress implementation in `pair_nequip`/`pair_allegro`
3. At the moment, no, but you can easily see the available model builders by looking in this folder: https://github.com/mir-group/nequip/tree/main/nequip/model. Most have docstrings describing what they do; for the default builders parameters with descriptions are also in `full.yaml`.

**Comment by youngjae142** [comment]
Hello @Linux-cpp-lisp !
First, thanks for your kind response!

So,
1. should I use [stress branch](https://github.com/mir-group/pair_nequip/tree/stress) of pair_nequip? 
I will let you know when I find an issue.
And yes, I need per-atom stress tensor calculation for lattice thermal conductivity calculation by the Green-Kubo method.
The [Green-Kubo method is implemented in LAMMPS](https://docs.lammps.org/compute_heat_flux.html) and it needs to run the [compute stress/atom](https://docs.lammps.org/compute_stress_atom.html) command.
The per-atom stress formalism is
S_ab = -m v_a v_b - W_{ab}
where the first term is the kinetic contribution (which is the universal form and so doesn't need to be implemented)
and the second term is the virial term which originates from interactions(force) between atoms.
I'm not sure but I suppose pair_nequip is a pairwise potential in practical.
So I believe the formalism of W_{ab} term only includes the first term (pairwise energy contribution) 
![image](https://user-images.githubusercontent.com/42454827/209062011-8dc1597d-a877-49b4-af6c-3d2a1534b351.png)
of the W_{ab} formalism described in [compute stress/atom](https://docs.lammps.org/compute_stress_atom.html) command.
I will test it with the current version of the stress branch.

2. Okay. I suppose what I need is StressForceOutput.

3. The model folder and full.yaml file are exactly what I needed. Thanks a lot!
